polynote-kernel/src/main/scala/polynote/kernel/lang/scal/ScalaSource.scala

```diff
@@ -274,12 +296,180 @@ class ScalaSource[G <: Global](
       unit
   }
 
+  // compiled unit but this time with all our fancy substitutions to help prevent serialization issues
+  lazy val compileUnitWithSubstitutions: Either[Throwable, global.CompilationUnit] = compileUnit.right.flatMap {
+    unit =>
+      Either.catchNonFatal {
+        val tree = unit.body
+
+        // ok, so case classes can't handle being put through the typer twice (and we can't roll back the changes that
+        // the typer does to them...) so, what we'll do here is copy any case classes we have in the tree before we type it.
+        // later, we'll sub the typed case class for the untyped case class so it can go through the typer again happily.
+        val caseClasses = (tree match {
+          case global.PackageDef(_, stats) =>
+            stats.collect {
+              case cls: global.ClassDef if cls.mods.isCase =>
+                cls.name.toString -> deepCopyTree(cls)
+            }
+        }).toMap
+        // we also need to keep track of any user-defined companion objects there might be, so that we can substitute them too.
+        val companionObjects = (tree match {
+          case global.PackageDef(_, stats) =>
+            stats.collect {
+              case o: global.ModuleDef if caseClasses.contains(o.name.toString) =>
+                o.name.toString -> deepCopyTree(o)
+            }
+        }).toMap
+
+        import global.Quasiquote
+
+        // type the tree so we can reify implicits and all that nice stuff.
+        val typedPkg = reporter.attempt {
+          val run = new global.Run()
+          global.globalPhase = run.namerPhase // make sure globalPhase matches run phase
+          run.namerPhase.asInstanceOf[global.GlobalPhase].apply(unit)
+          global.globalPhase = run.typerPhase // make sure globalPhase matches run phase
+          run.typerPhase.asInstanceOf[global.GlobalPhase].apply(unit)
+          global.exitingTyper(unit.body)
+        } match {
+          case Right(t) => t
+          case Left(err) => throw err
+        }
+
+        // now that we've typed it all up we can look for references to previous cells.
+        val prevCellNames = previousSources.map(_.compiledModule.right.get.asModule.name)
+        val usedIdents = scala.collection.mutable.HashMap.empty[String, global.Tree]
+        val cellRefFinder = new global.Transformer {
```

Comment:
I think you can use a `Traverser` here instead, since you're not actually transforming anything. Also `typedPkg.collect` might achieve the same thing (not sure though, I could be confusing it with spark expressions)

Reply:
Oh very cool, that is pretty mught what `Tree.collect` does! 

https://github.com/scala/scala/blob/2.11.x/src/reflect/scala/reflect/internal/Trees.scala#L1654-L1660